core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java

```diff
@@ -138,20 +138,16 @@ protected boolean isCaseSensitive() {
   }
 
   protected PartitionSpec dataSpec() {
-    Set<PartitionSpec> specs = dataSpecs();
+    Set<PartitionSpec> specs =
+        newDataFilesBySpec.keySet().stream().map(this::spec).collect(ImmutableSet.toImmutableSet());
```

Comment:
This seems a bit over-complicated. At the end of this method, there will be just one spec, so why look up all the specs by ID and then fail? I think this should look more like this:

```java
  Set<Integer> specs = newDataFilesBySpec.keySet();
  Preconditions...
  return spec(Iterables.getOnlyElement(specs));
```